src/ngraph/runtime/plaidml/plaidml_builder.hpp

```diff
@@ -0,0 +1,391 @@
+/*******************************************************************************
+* Copyright 2018 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include <plaidml/plaidml++.h>
+
+#include <list>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <utility>
+
+#include "ngraph/runtime/plaidml/plaidml_config.hpp"
+
+// Utilities for constructing PlaidML functions.
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace plaidml
+        {
+            namespace builder
+            {
+                class BinaryContraction;
+                class ContractionInput;
+                class ContractionOutput;
+                class Elementwise;
+                class Function;
+                class Input;
+                class Output;
+                class Statement;
+                class TernaryContraction;
+                class UnaryContraction;
+            }
+        }
+    }
+}
+
+// Function provides a fluent interface for the construction of the text form of PlaidML functions.
+// It's intended to be simpler to use and produce simpler code than using direct string construction.
+//
+// N.B. It's possible to capture the intermediate pieces as they're being added to a function
+// (e.g. in order to directly code loops which call methods on them), but it's important to
+// remember that what's returned are references, not objects; the caller is responsible for
+// instantiating a Function instance and keeping it alive as long as there are any outstanding
+// references to its constituent pieces.
+class ngraph::runtime::plaidml::builder::Function final
+{
+public:
+    Function(const std::string& name, bool debug);
+    Function(const Function&) = delete;
+    Function& operator=(const Function&) = delete;
+    Function(Function&&) = default;
+    Function& operator=(Function&&) = default;
+
+    // Builds the final string form of the function.
+    std::string to_string() const;
+
+    // Finalizes a function, transforming it into a PlaidML function application object.
+    vertexai::plaidml::application finalize() const;
+
+    // Adds an input to the function.
+    Function& add(Input input) &;
```

Comment:
What is the problem being solved with returning `&` and `&&`? I haven't used this before.

Reply:
The idea was to make it easy to use `this` as either an lvalue and rvalue-reference, so that you could do either `MyObject foo; foo.bar().baz(); return foo.finalize()` or `return MyObject{}.bar().baz().finalize()`, depending on what reads best at the call site.